source/UF/Truncations.lagda

```diff
@@ -0,0 +1,435 @@
+Ian Ray, 07/23/2024
+
+Using records we define the general truncation of a type which will include
+a constructor, an induction principle and a computation rule
+(up to identification). We then proceed to develop some machinery derived from
+the induction principle and explore relationships, closure properties and finally
+characterize the identity type of truncations. Note that we explicitly include
+the assumption of univalence for the characterization of idenity types but not
+globally as many important properties hold in the absence of univalence.
+
+\begin{code}
+
+{-# OPTIONS --safe --without-K #-}
+
+open import UF.FunExt
+open import UF.PropTrunc 
+
+module UF.Truncations (fe : Fun-Ext)
+                      (pt : propositional-truncations-exist)
+                       where
+
+open import MLTT.Spartan
+open import UF.Base
+open import UF.Embeddings
+open import UF.Equiv
+open import UF.EquivalenceExamples
+open import UF.Equiv-FunExt
+open import UF.H-Levels fe
+open import UF.IdentitySystems
+open import UF.Retracts
+open import UF.Sets
+open import UF.Singleton-Properties
+open import UF.Subsingletons
+open import UF.Subsingletons-FunExt
+open import UF.Subsingletons-Properties
+open import UF.Univalence
+open import UF.UA-FunExt
+open import UF.Yoneda
+open import Naturals.Addition renaming (_+_ to _+'_)
+open import Naturals.Order
+
+\end{code}
+
+We define the notion of a n-truncation using record types.
+
+\begin{code}
+
+record H-level-truncations-exist : 𝓤ω where
+ field
+  ∥_∥[_] : {𝓤 : Universe} → 𝓤 ̇ → ℕ → 𝓤 ̇
+  ∥∥ₙ-hlevel : {𝓤 : Universe} {X : 𝓤 ̇ } (n : ℕ) → ∥ X ∥[ n ] is-of-hlevel n
+  ∣_∣[_] :  {𝓤 : Universe} {X : 𝓤 ̇ } → X → (n : ℕ) → ∥ X ∥[ n ]
+  ∥∥ₙ-ind : {X : 𝓤 ̇ } {n : ℕ} {P : ∥ X ∥[ n ] → 𝓥 ̇}
+          → ((s : ∥ X ∥[ n ]) → (P s) is-of-hlevel n)
+          → ((x : X) → P (∣ x ∣[ n ]))
+          → (s : ∥ X ∥[ n ]) → P s
+  ∥∥ₙ-ind-comp : {X : 𝓤 ̇ } {n : ℕ} {P : ∥ X ∥[ n ] → 𝓥 ̇ }
+               → (m : (s : ∥ X ∥[ n ]) → (P s) is-of-hlevel n)
+               → (g : (x : X) → P (∣ x ∣[ n ]))
+               → (x : X) → ∥∥ₙ-ind m g (∣ x ∣[ n ]) ＝ g x
+ infix 0 ∥_∥[_]
+ infix 0 ∣_∣[_]
+
+\end{code}
+
+We now add some some machinery that will prove useful: truncation recursion
+and uniqueness, induction/recursion for two arguments and all associated
+computation rules.
+
+\begin{code}
+
+ ∥∥ₙ-rec : {X : 𝓤 ̇ } {Y : 𝓥 ̇ } {n : ℕ}
+         → Y is-of-hlevel n
+         → (X → Y)
+         → ∥ X ∥[ n ] → Y
+ ∥∥ₙ-rec {𝓤} {𝓥} {X} {Y} {n} Y-h-level f s =
```

Comment:
You don't need any of these arguments `{𝓤} {𝓥} {X} {Y} {n}` and it can be just be one line.